File: opentime.js
// Imports
var request = require("request");
var MongoClient = require('mongodb').MongoClient;
var assert = require('assert');
var querystring = require('querystring');
var cheerio = require('cheerio');

// var URL_OF_LIBRARY_API = "https://ajax.googleapis.com/ajax/services/feed/load?v=1.0\u0026amp;num=-1\u0026amp;q=http://www.feed43.com/2343747281410148.xml";
var URL_OF_LIBRARY_API = "http://52.44.150.99/users/1/web_requests/15/library.json";

// Libraries
var LIST_OF_LIBRARIES = [
    "kf - Academic Success Centre, Koffler Centre",
    "koffler - Academic Success Centre, Koffler Centre",
    "astronomy - Astronomy \u0026amp; Astrophysics Library",
    "chem - Chemistry Library (A D Allen)",
    "allen - Chemistry Library (A D Allen)",
    "art - Department of Art Library",
    "engineering - Engineering \u0026amp; Computer Science Library",
    "aerospace - Engineering \u0026amp; Computer Science Library - Aerospace Resource Centre",
    "medicine - Family \u0026amp; Community Medicine Library",
    "newman - Industrial Relations and Human Resources Library (Newman)",
    "law - Law Library (Bora Laskin)",
    "map - Map and Data Library: Collection Access",
    "music - Music Library",
    "new - New College Library (Ivey)",
    "petro - Petro Jacyk Central \u0026amp; East European Resource Centre",
    "regis - Regis College Library",
    "hk - Richard Charles Lee Canada-Hong Kong Library",
    "smc - St. Michael’s College - John M. Kelly Library",
    "kelly - St. Michael’s College - John M. Kelly Library",
    "trinity -  Trinity College Library (John W Graham Library)",
    "emmanuel - Victoria University - Emmanuel College Library",
    "ba - Bahen Centre o(*////▽////*)q",
    "rom - Royal Ontario Museum Library \u0026amp; Archives",
    "oi - OISE Library",
    "gerstein - Gerstein Science Information Centre",
    "ej - E J Pratt Library",
    "rb - Robarts Library",
];

/* All values in "DICT_OF_LIBRARIES" must also be in "LIST_OF_LIBRARIES" */
var DICT_OF_LIBRARIES = {
    "kf": "Academic Success Centre, Koffler Centre",
    "koffler": "Academic Success Centre, Koffler Centre",
    "academic": "Academic Success Centre, Koffler Centre",
    "astronomy": "Astronomy \u0026amp; Astrophysics Library",
    "astrophysics": "Astronomy \u0026amp; Astrophysics Library",
    "chem": "Chemistry Library (A D Allen)",
    "chemistry": "Chemistry Library (A D Allen)",
    "allen": "Chemistry Library (A D Allen)",
    "art": "Department of Art Library",
    "engineering": "Engineering \u0026amp; Computer Science Library",
    "aerospace": "Engineering \u0026amp; Computer Science Library - Aerospace Resource Centre",
    "family": "Family \u0026amp; Community Medicine Library",
    "medicine": "Family \u0026amp; Community Medicine Library",
    "newman": "Industrial Relations and Human Resources Library (Newman)",
    "law": "Law Library (Bora Laskin)",
    "map": "Map and Data Library: Collection Access",
    "music": "Music Library",
    "new": "New College Library (Ivey)",
    "ivey": "New College Library (Ivey)",
    "petro": "Petro Jacyk Central \u0026amp; East European Resource Centre",
    "regis": "Regis College Library",
    "richard": "Richard Charles Lee Canada-Hong Kong Library",
    "hk": "Richard Charles Lee Canada-Hong Kong Library",
    "robarts": "Robarts Library",
    "rl": "Robarts Library",
    "rom": "Royal Ontario Museum Library \u0026amp; Archives",
    "rare": "Thomas Fisher Rare Book Library",
    "smc": "St. Michael’s College - John M. Kelly Library",
    "trinity": "Trinity College Library (John W Graham Library)",
    "oi": "OISE Library",
    "kelly": "St. Michael’s College - John M. Kelly Library",
    "rb": "Robarts Library",
    "gerstein": "Gerstein Science Information Centre",
    "ej": "Victoria University - E J Pratt Library",
    "emmanuel": "Victoria University - Emmanuel College Library"
};

// Get all libraries -- can be simplified
function getSuggestedLibraries() {
    var suggested_libraries = '';
    for (var i = 0; i < LIST_OF_LIBRARIES.length; i++) {
        suggested_libraries += LIST_OF_LIBRARIES[i];
        suggested_libraries += '\n\n';
    }
    return suggested_libraries;
}

// Get output according to input
exports.library = function(query, callback) {
    var prefix = query.split(/\s+/)[0]; // "lib" in "lib rb"
    var libInput = query.split(/\s+/)[1]; // "rb" in "lib rb"
    var newInput = '';
    if (libInput == undefined) { // If only one token
        callback("你要找哪些图书馆呢？\n" + getSuggestedLibraries());
        return;
    }

    /* Some surprise */
    if(libInput == "ba"){
        callback("24/7/365, 程序员不用休息哒（¯﹃¯）");
        return;
    }

    if (DICT_OF_LIBRARIES[libInput] == undefined) { //输入的内容无法在dict中找到，直接查找不转换
            // console.log("DEBUG: NOT FOUND IN DICT");
            callback("抱歉哦,小助手找不到你输入的图书馆,你是不是要找以下的图书馆呢?>_<\n " + getSuggestedLibraries() + "请输入图书馆简称哦 (e.g.: library kf) _(:3 」∠)_ ");
            return;
    } else { //输入的内容可以在DICT找得到，先转换成全称
        newInput = DICT_OF_LIBRARIES[libInput];
        var URL_BASE = 'http://resource.library.utoronto.ca/hours/';
        var options = {
          url: URL_BASE,
          headers: {
            'User-Agent': 'request' // need a user agent to scrape page correctly.
          }
        };
        request(options, function (error, response, html) {
          if (!error && response.statusCode == 200) {
            // console.log(html);
            var $ = cheerio.load(html);
            var libLst = [];

            $('div.library-row').each(function(i, element){
                var name = $(this).children('div.library').children('h2').text().trim();
                var time = $(this).children('div.library-hours').text().trim() || "unavailable";
                var info = $(this).children('div.library').children('div.librarynotes').text().trim() || "";
                var url = URL_BASE + $(this).children('div.library-month').children('a').attr('href') || "";
                libLst.push({
                    "name": name,
                    "time": time,
                    "info": info,
                    "url": url
                });
            });

            /*
              Given libLst, match with newInput and return stats.
             */
            for (var i = 0; i < libLst.length; i++) {
                var lib_title = libLst[i]["name"];
                var lib_notes = libLst[i]["info"] ? "\n提示：" + libLst[i]["info"] : "";
                if (lib_title == newInput) {
                    callback("Open time for " + lib_title.split("&amp;").join("") + ":\n" +  libLst[i]["time"] + lib_notes + "\n查看未来几天的开放时间，看这里哦：\n" + libLst[i]["url"]);
                    return;
                }
            }
            callback("找不到该图书馆。");
            return;
          };
        });
    }
}

File: opentime.py
# -*- coding: utf-8 -*-
from bs4 import BeautifulSoup
#import urllib.request
import urllib

DICT_OF_LIBRARIES = {
    #"kf": "Career Resource Library, Koffler Centre",
    #"koffler": "Career Resource Library, Koffler Centre",
    "astronomy": "Astronomy & Astrophysics Library",
    "astrophysics": "Astronomy & Astrophysics Library",
    "chem": "Chemistry Library (A D Allen)",
    "chemistry": "Chemistry Library (A D Allen)",
    "allen": "Chemistry Library (A D Allen)",
    "art": "Department of Art Library",
    "engineering": "Engineering & Computer Science Library",
    "aerospace": "Engineering & Computer Science Library - Aerospace Resource Centre",
    "family": "Family & Community Medicine Library",
    "medicine": "Family & Community Medicine Library",
    "newman": "Industrial Relations and Human Resources Library (Newman)",
    "law": "Law Library (Bora Laskin)",
    #"map": "Map and Data Library",
    "music": "Music Library",
    "new": "New College Library (Ivey)",
    "ivey": "New College Library (Ivey)",
    "petro": "Petro Jacyk Central & East European Resource Centre",
    "regis": "Regis College Library",
    "richard": "Richard Charles Lee Canada-Hong Kong Library",
    "hk": "Richard Charles Lee Canada-Hong Kong Library",
    "robarts": "Robarts Library",
    "rl": "Robarts Library",
    "rom": "Royal Ontario Museum Library & Archives",
    "rare": "Thomas Fisher Rare Book Library",
    "smc": "St. Michael’s College - John M. Kelly Library",
    "trinity": "Trinity College Library (John W Graham Library)",
    "oi": "OISE Library",
    "kelly": "St. Michael’s College - John M. Kelly Library",
    "rb": "Robarts Library",
    "gerstein": "Gerstein Science Information Centre",
    "ej": "Victoria University - E J Pratt Library",
    "emmanuel": "Victoria University - Emmanuel College Library"
};

LIST_OF_LIBRARIES = [
    #"kf - Career Resource Library, Koffler Centre",
    #"koffler - Career Resource Library, Koffler Centre",
    "astronomy - Astronomy \u0026amp; Astrophysics Library",
    "chem - Chemistry Library (A D Allen)",
    "allen - Chemistry Library (A D Allen)",
    "art - Department of Art Library",
    "engineering - Engineering \u0026amp; Computer Science Library",
    "aerospace - Engineering \u0026amp; Computer Science Library - Aerospace Resource Centre",
    "medicine - Family \u0026amp; Community Medicine Library",
    "newman - Industrial Relations and Human Resources Library (Newman)",
    "law - Law Library (Bora Laskin)",
    #"map - Map and Data Library",
    "music - Music Library",
    "new - New College Library (Ivey)",
    "petro - Petro Jacyk Central \u0026amp; East European Resource Centre",
    "regis - Regis College Library",
    "hk - Richard Charles Lee Canada-Hong Kong Library",
    "smc - St. Michael's College - John M. Kelly Library",
    "kelly - St. Michael's College - John M. Kelly Library",
    "trinity -  Trinity College Library (John W Graham Library)",
    "emmanuel - Victoria University - Emmanuel College Library",
    "ba - Bahen Centre",
    "rom - Royal Ontario Museum Library \u0026amp; Archives",
    "oi - OISE Library",
    "gerstein - Gerstein Science Information Centre",
    "ej - E J Pratt Library",
    "rb - Robarts Library",
];

def getOpentime(second):
    if second == "ba":
        #return "24/7/365, 程序员不用休息哒（¯﹃¯）<br/>"
        return "Everyday!!! Computer scientist do not need a break~<br/>"
        #return 'pass'

    # If the library entered is not in the DICT, then ask the user to enter library again
    if second not in DICT_OF_LIBRARIES:
        #return "抱歉哦,小助手找不到你输入的图书馆,你是不是要找以下的图书馆呢?>_<<br/>" + getSuggestedLibraries() + "请输入图书馆简称哦 (e.g.: library kf) _(:3 」∠)_ "
        return "Sorry, the library you entered is not valid.<br/> Please re-enter the  library you are looking for using the abbreviation. (i.e. lib rb)<br/>" + getSuggestedLibraries()
    # If the library is in the DICT, then parser the web for information
    else:
        lib_dic = getInfo()
        name = DICT_OF_LIBRARIES[second].decode('utf-8').encode('ascii', 'ignore')
        if name not in lib_dic.keys():
            return name + " Opps! some error occurs"
        hour = lib_dic[name][0]
        website = lib_dic[name][1]

        if hour == 0:
            #return 'pass'
            return " Open Time for {}: <br/>Please contact library for hours.<br/>Library Information:<br/><a href =\"{}\">{}</a>".format(name, website, website)
        else:
            #return 'pass'
            return " Open Time for {}: <br/>{}<br/>Library Information:<br/><a href =\"{}\">{}</a>".format(name, hour, website, website)

#==========helper function==========
def getInfo():
    url = 'https://onesearch.library.utoronto.ca/visit'
    # r = urllib.request.urlopen(url).read()
    r = urllib.urlopen(url).read()
    soup = BeautifulSoup(r, 'html.parser')
    
    count = 0
    lib_dic = {}
    all_libs = soup.select('div.block-content div.views-row')

    for lib in all_libs:
        name = lib.select('span h2 a')[0].get_text().encode('ascii', 'ignore')
        hour_container = lib.select('div.featured-library-home h3 a')
        if len(hour_container) < 1:
            hour = 0
        else:
            hour = hour_container[0]['href'].encode('ascii', 'ignore')
        website = url + lib.select('span h2 a')[0]['href'].encode('ascii', 'ignore')
        lib_dic[name] = [hour, website]

    return lib_dic

def getSuggestedLibraries():
    count = 0
    for item in LIST_OF_LIBRARIES:
        count += 1
    #print count
    return '<br/>'.join(LIST_OF_LIBRARIES)

if __name__ == '__main__':
    for key in DICT_OF_LIBRARIES:
        print getOpentime(key)
    #getOpentime('rb')